ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs

```diff
@@ -0,0 +1,92 @@
+       ( StorageDecodingError (..)
+       , UserSecret (..)
+       , Version (..)
+       , eskToKey
+       , getUserSecretKeys
+       , usFilePath
+       , usWallets
+       , usVersion
+       ) where
+
+import Control.Lens.TH (makeLenses)
+import Data.Aeson
+  (FromJSON, ToJSON, Value(..), toJSON, parseJSON, withText, withObject, (.:), object, (.=))
+import Data.Default
+import Data.Map (Map)
+import Formatting (bprint, build)
+
+import Pos.Binary.Class (serialize', decodeFull')
+import Pos.Core (AddressHash, addressHash)
+import Pos.Crypto (EncryptedSecretKey(..), PublicKey, encToPublic)
+
+import qualified Data.Text.Buildable
+import qualified Data.Map as Map (empty, fromList, elems)
+import qualified Serokell.Util.Base16 as S (encode, decode)
+
+newtype Version= Version { getVersion :: Int }
+  deriving (Show, Eq)
+
+-- | UserSecret data containing wallet keys that gets saved to file
+data UserSecret
+  = UserSecret
+  { _usWallets  :: Map (AddressHash PublicKey) EncryptedSecretKey
+  , _usVersion  :: !Version
+  , _usFilePath :: !FilePath
+  } deriving (Show)
+
+makeLenses ''UserSecret
+
+instance Default UserSecret where
+  def = UserSecret
+      { _usWallets  = Map.empty
+      , _usVersion  = Version 1
+      , _usFilePath = mempty
+      }
+
+newtype StorageDecodingError = StorageDecodingError Text
+    deriving (Show)
+
+instance Exception StorageDecodingError
+
+instance Buildable StorageDecodingError where
+    build (StorageDecodingError msg) =
+        "Failed to decode stored json: " <> bprint build msg
+
+instance ToJSON EncryptedSecretKey where
+  toJSON = String . S.encode . serialize'
+
+instance FromJSON EncryptedSecretKey where
+  parseJSON  = withText "EncryptedSecretKey" $ \encryptedText ->
+                let textFail     = fail . toString
+                    eitherDecode = either textFail pure . decodeFull'
+                in either textFail eitherDecode $ S.decode encryptedText
+
+instance ToJSON UserSecret where
+  toJSON us =
+    object [ "keys"    .= getUserSecretKeys us
+           , "version" .= (getVersion $ _usVersion us)
+           ]
+
+instance FromJSON UserSecret where
+  parseJSON = withObject "UserSecret" $ \v ->
+                let pEncMap  = esksToMap <$> v .: "keys"
+                    pVersion = Version   <$> v .: "version"
+                in UserSecret <$> pEncMap <*> pVersion <*> pure mempty
+
+{-------------------------------------------------------------------------------
+  Utilities
+-------------------------------------------------------------------------------}
+
+esksToMap :: [EncryptedSecretKey] -> Map (AddressHash PublicKey) EncryptedSecretKey
+esksToMap esks =
+    esks <&> (\encKey -> (eskToKey encKey, encKey))
+          & Map.fromList
+
+eskToKey :: EncryptedSecretKey -> AddressHash PublicKey
+eskToKey = addressHash . encToPublic
+
+getUserSecretKeys :: UserSecret -> [EncryptedSecretKey]
+getUserSecretKeys us = Map.elems (us ^. usWallets)
```

Comment:
Actually, `eskToKey` and `getUserSecretKeys` look a lot like functions that could be put in a `Util` module.

Reply:
Good call

Reply:
Turns we can't have `getUserSecretKeys` in a `Util` module due to `cyclic dependence issue` between the 2 modules. I added some other functions there though.